sql/sql_table.cc

```diff
@@ -775,7 +775,23 @@ bool mysql_write_frm(ALTER_PARTITION_PARAM_TYPE *lpt, uint flags)
   strxmov(shadow_frm_name, shadow_path, reg_ext, NullS);
   if (flags & WFRM_WRITE_SHADOW)
   {
-    if (mysql_prepare_create_table(lpt->thd, lpt->create_info, lpt->alter_info,
+    /*
+      It is possible mysql_prepare_create_table was already called in our
+      create/alter_info context and we don't need to call it again. That is, if
+      in the context of `ALTER TABLE ... CONVERT PARTITION .. TO TABLE` then
+      mysql_prepare_create_table would have already been called through a prior
+      invocation of mysql_write_frm with flag MFRM_WRITE_CONVERTED_TO.
+
+      We check if mysql_prepare_create_table was already called by checking if
+      the alter_info struct's keys have already been initialized, which is done
+      in mysql_prepare_create_table_finalize (via function init_key_info()).
+    */
+    bool prepare_create_table_already_called= false;
```

Comment:
Consider compactifying this a little:
```
auto &key_list= lpt->alter_info->key_list;
bool prepare_create_table_already_called= !key_list.is_empty() && key_list.head()->length > 0;
```

Reply:
Thanks @FooBarrior !

Though I think I'm still going to implement

> I'd propose to pass an extra (optional, maybe) argument that hints a possible second time entrance, so that we could distinguish uninitialized key list and this.

as I think that is more future-proof (and will simplify the expression even more :smile:)

Reply:
You opposed to this so much just in the other comment. What made you change your mind?

Reply:
yeah, and i thought about adding that argument to `mysql_prepare_create_table`. no pasa nada, reading your new commit